include/revng/Pipeline/Analysis.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/ArrayRef.h"
+
+#include "revng/Pipeline/Invokable.h"
+#include "revng/Pipeline/Pipe.h"
+
+namespace pipeline {
+
+template<typename Analysis>
+class AnalysisWrapperImpl;
+
+class AnalysisWrapperBase : public InvokableWrapperBase {
+private:
+  std::string CustomName;
+
+public:
+  template<typename Analysis>
+  using ImplType = AnalysisWrapperImpl<Analysis>;
+
+public:
+  virtual ~AnalysisWrapperBase() override = default;
+  virtual llvm::ArrayRef<Kind *>
+  getAcceptedKinds(size_t ContainerIndex) const = 0;
+
+  const std::string &getCustomName() const { return CustomName; }
+  void setCustomName(std::string NewName) { CustomName = std::move(NewName); }
+};
+
+template<typename Analysis>
+class AnalysisWrapperImpl : public AnalysisWrapperBase {
+private:
+  InvokableWrapperImpl<Analysis> Invokable;
+
+public:
+  AnalysisWrapperImpl(Analysis ActualPipe,
+                      std::vector<std::string> RunningContainersNames) :
+    Invokable(std::move(ActualPipe), std::move(RunningContainersNames)) {}
+
+  ~AnalysisWrapperImpl() override = default;
+
+public:
+  llvm::ArrayRef<Kind *>
+  getAcceptedKinds(size_t ContainerIndex) const override {
+    return Invokable.getPipe().AcceptedKinds[ContainerIndex];
+  }
+
+  void dump(std::ostream &OS, size_t Indentation) const override {
+    Invokable.dump(OS, Indentation);
+  }
+
+  void print(const Context &Ctx,
+             llvm::raw_ostream &OS,
+             size_t Indentation) const override {
+    Invokable.print(Ctx, OS, Indentation);
+  }
+
+  void run(Context &Ctx, ContainerSet &Containers) override {
+    Invokable.run(Ctx, Containers);
+  }
+
+  std::vector<std::string> getRunningContainersNames() const override {
+    return Invokable.getRunningContainersNames();
+  }
+  std::string getName() const override { return Invokable.getName(); }
```

Comment:
What's the difference between `getName` and `getCustomName`?
Is this naming reflected in the pipeline?

Reply:
"name" is the "type" in the yaml, "customName" is the "name" in the yaml

Reply:
OK, let's make these coherent. No need to deviate.

Reply:
the issue is that the name is as well the name in the class declaration, like in pipes

Reply:
Yes, but pipe names cannot be customized.
I'd go for `getName`+`getTypeName`.